[RFC] Fix GCC 7 issues

jamessan opened the pull request:
Address known issues with the upcoming GCC 7 stable release.

- [x] `-Wimplicit-fallthrough`
- [x] `-Wpointer-compare`

Closes #6512

jamessan committed:
func_attr: Support "noreturn" attribute

jamessan committed:
os_unix: Mark mch_exit as NORETURN

jamessan committed:
ex_cmds: Use NULL, not NUL, to check for strrchr failure

jamessan committed:
*: Comment intentional fallthroughs

Falling through a switch case should be commented so it's clear that
behavior is intentional.

jamessan commented:
All the blatantly obvious intentional fall through cases have been commented.  What's left is a mix of cases that need more analysis and cases that are [false positives](https://bugs.debian.org/860263):

```
[117/250] Building C object src/nvim/CMakeFiles/nvim.dir/ex_getln.c.o
/home/jamessan/src/github.com/neovim/src/nvim/ex_getln.c: In function 'command_line_handle_key':
/home/jamessan/src/github.com/neovim/src/nvim/ex_getln.c:1260:8: warning: this statement may fall through [-Wimplicit-fallthrough=]
     if (s->xpc.xp_numfiles > 0) {
        ^
/home/jamessan/src/github.com/neovim/src/nvim/ex_getln.c:1268:3: note: here
   case K_UP:
   ^~~~
[143/250] Building C object src/nvim/CMakeFiles/nvim.dir/eval/typval.c.o
/home/jamessan/src/github.com/neovim/src/nvim/eval/typval.c: In function 'tv_free':
/home/jamessan/src/github.com/neovim/src/nvim/eval/typval.c:1864:9: warning: this statement may fall through [-Wimplicit-fallthrough=]
         func_unref(tv->vval.v_string);
         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/jamessan/src/github.com/neovim/src/nvim/eval/typval.c:1867:7: note: here
       case VAR_STRING: {
       ^~~~
[151/250] Building C object src/nvim/CMakeFiles/nvim.dir/edit.c.o
/home/jamessan/src/github.com/neovim/src/nvim/edit.c: In function 'quote_meta':
/home/jamessan/src/github.com/neovim/src/nvim/edit.c:4914:11: warning: this statement may fall through [-Wimplicit-fallthrough=]
           || ctrl_x_mode == CTRL_X_THESAURUS)
           ^
/home/jamessan/src/github.com/neovim/src/nvim/edit.c:4916:5: note: here
     case '~':
     ^~~~
/home/jamessan/src/github.com/neovim/src/nvim/edit.c:4917:10: warning: this statement may fall through [-Wimplicit-fallthrough=]
       if (!p_magic)             /* quote these only if magic is set */
          ^
/home/jamessan/src/github.com/neovim/src/nvim/edit.c:4919:5: note: here
     case '\\':
     ^~~~
/home/jamessan/src/github.com/neovim/src/nvim/edit.c:4921:11: warning: this statement may fall through [-Wimplicit-fallthrough=]
           || ctrl_x_mode == CTRL_X_THESAURUS)
           ^
/home/jamessan/src/github.com/neovim/src/nvim/edit.c:4923:5: note: here
     case '^':                   /* currently it's not needed. */
     ^~~~
[190/250] Building C object src/nvim/CMakeFiles/nvim.dir/normal.c.o
/home/jamessan/src/github.com/neovim/src/nvim/normal.c: In function 'do_pending_operator':
/home/jamessan/src/github.com/neovim/src/nvim/normal.c:1856:10: warning: this statement may fall through [-Wimplicit-fallthrough=]
       if (vim_strchr(p_cpo, CPO_FILTER) != NULL) {
          ^
/home/jamessan/src/github.com/neovim/src/nvim/normal.c:1862:5: note: here
     case OP_INDENT:
     ^~~~
[201/250] Building C object src/nvim/CMakeFiles/nvim.dir/shada.c.o
/home/jamessan/src/github.com/neovim/src/nvim/shada.c: In function 'shada_read_when_writing':
/home/jamessan/src/github.com/neovim/src/nvim/shada.c:2046:13: warning: this statement may fall through [-Wimplicit-fallthrough=]
         ret = kSDWriteReadNotShada;
         ~~~~^~~~~~~~~~~~~~~~~~~~~~
/home/jamessan/src/github.com/neovim/src/nvim/shada.c:2049:7: note: here
       case kSDReadStatusReadError: {
       ^~~~
[212/250] Building C object src/nvim/CMakeFiles/nvim.dir/regexp.c.o
In file included from /home/jamessan/src/github.com/neovim/src/nvim/regexp.c:7031:0:
../src/nvim/regexp_nfa.c: In function 'nfa_recognize_char_class':
../src/nvim/regexp_nfa.c:630:19: warning: this statement may fall through [-Wimplicit-fallthrough=]
         } else if (*(p + 2) == '7') {
                   ^
../src/nvim/regexp_nfa.c:634:7: note: here
       case 'a':
       ^~~~
../src/nvim/regexp_nfa.c:638:19: warning: this statement may fall through [-Wimplicit-fallthrough=]
         } else if (*(p + 2) == 'f') {
                   ^
../src/nvim/regexp_nfa.c:642:7: note: here
       case 'A':
       ^~~~
```

justinmk commented on src/nvim/eval.c:
```diff
@@ -9595,13 +9595,15 @@ static void f_getcwd(typval_T *argvars, typval_T *rettv, FunPtr fptr)
         break;
       }
-    case kCdScopeTab:         // FALLTHROUGH
+      // fallthrough
```
Does it work with uppercase? I'd prefer that, so it looks like a TODO .

ZyX-I replied:
I do not see a reason to shout here, fallthrough is not something which requires action. It is also not an abbreviation of some kind.

justinmk replied:
Well, it's comment that has special meaning, so making it look like a normal comment is misleading. But I don't feel strongly about it.

justinmk replied:
So now that we have the macro, should we not use it in all cases?

jamessan replied:
I read @ZyX-I's [comment](https://github.com/neovim/neovim/issues/6512#issuecomment-293977487) to mean "use comments except where an attribute is needed", but I could have misinterpreted it.  I don't have a strong opinion either way.

ZyX-I replied:
@jamessan My comment indeed means that, but I would not mind macros in place of comments as well.

jamessan commented:
According to an upstream [bug], it looks like they don't have any plans to fix the handling of

```c
case 1: {
  foo = 0;
  //fallthrough
}
```

Apparently, a `}` is too hard to propagate information across, so they don't know about the comment after parsing that...

Guess the "difficult" cases will need to use the macro, as @ZyX-I [suggested].

[bug]: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=77817#c11
[suggested]: https://github.com/neovim/neovim/issues/6512#issuecomment-293958257

jamessan committed:
Detect support for and use -Wimplicit-fallthrough

jamessan committed:
*: Use __attribute__((fallthrough)) where comments aren't supported

Although GCC now detects possibly unintentional fall through, there
rules around which the comments are detected are rather strict.  In
cases where a comment isn't detected, upstream [recommends] using their
fallthrough attribute.

[recommends]: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=77817#c11

jamessan commented:
Still a few parts of the code to review, but all the infrastructure is in place now.

jamessan committed:
regexp_nfa: Fix invalid fallthrough in character class detection

When the end character in a range matches a different standard range
(e.g., [0-z]), the range would be incorrectly detected as the class of
the end character (CLASS_az).

Instead of using a fallthrough, immediately FAIL when the end character
doesn't match the expected range.

jamessan committed:
lint

jamessan commented:
I also sent 0e8e48d [upstream], so we may want to wait for that to annotate the patch.  Otherwise, this should be ready.

[upstream]: https://github.com/vim/vim/pull/1703

justinmk commented on CMakeLists.txt:
```diff
@@ -221,6 +221,13 @@ else()
   add_definitions(-Wall -Wextra -pedantic -Wno-unused-parameter
     -Wstrict-prototypes -std=gnu99)
 
+  if(CMAKE_COMPILER_IS_GNUCC)
```
why not check the flag for all compilers? clang may add it eventually.

jamessan replied:
We can't check for something if we don't know what it will look like.  For example, GCC's "ignore UBSAN" looks like `__attribute__((no_sanitize_undefined))` but clang's looks like `__attribute__((no_sanitize("undefined")))`.

I suppose this check should be for the presence of the attribute, rather than the compile flag, though.

jamessan replied:
Added detection of `-Wimplicit-fallthrough` so we can explicitly add it to the compile flags, and changed the attribute macro to be defined based on `#if __has_attribute(fallthrough)`.

jamessan commented:
Hmm, older clang is having problems with the `__has_attribute()` stuff, even though I found it in their documentation first (later in GCC's).

jamessan commented:
Ok, clang older than 3.6 reported true if _any_ backend supported the attribute.  In 3.6 and later, `__has_attribute(foo)` is only true for GNU-style attributes, which is what we want.  This should fix the QB builds.

jamessan committed:
oldtests: Fix spelling of testname in s:flaky

oni-link commented:
When playing around with a few [attributes](https://clang.llvm.org/docs/AttributeReference.html#returns-nonnull-gnu-returns-nonnull) for clang-5.0,
```diff

diff --git a/src/nvim/func_attr.h b/src/nvim/func_attr.h
index cc94a41f8..4c2db79bf 100644
--- a/src/nvim/func_attr.h
+++ b/src/nvim/func_attr.h
@@ -114,6 +114,9 @@
 #  define REAL_FATTR_NORETURN __attribute__((noreturn))
 
 #  ifdef __clang__
+#    define REAL_FATTR_ALLOC_SIZE(x) __attribute__((alloc_size(x)))
+#    define REAL_FATTR_ALLOC_SIZE_PROD(x, y) __attribute__((alloc_size(x, y)))
+#    define REAL_FATTR_NONNULL_RET __attribute__((returns_nonnull))
 // clang only
 #  elif defined(__INTEL_COMPILER)
 // intel only
```
the following was found,
```bash
/home/oni-link/git/neovim/src/nvim/path.c:441:5: warning: null returned from function that requires a non-null return value [-Wnonnull]
    return NULL;
    ^      ~~~~
```

This was not detected with gcc-7.1.

justinmk commented:
@oni-link see #6735

justinmk commented:
https://github.com/vim/vim/pull/1703 could take a while since Bram is busy. I think we should go ahead with this.

jamessan commented:
Ok. I'll just add it to the vim 8 patches PR when he merges.
